### DISCUSSION TITLE
Calling V from C section too small

### DISCUSSION THREAD

**Discussion by nicolasventer** [vlang/v#15754] [Ideas]
Hello,  
I want to call V from C. But since the documentation is really short, I don't know how I can do it.  
This is what I've tried:  
*In a.v:*
```v
module a

[export: 'my_custom_c_name']
fn foo() {
	println("foo")
}
```
*In b.c:*
```c
#include "a.v"

int main()
{
	my_custom_c_name();
}
```
And in cmd, I execute ``gcc b.c`` and I got:
```bash
> gcc b.c
In file included from b.c:1:
a.v:1:1: error: unknown type name 'module'
    1 | module a
      | ^~~~~~
a.v:3:2: error: 'export' undeclared here (not in a function)       
    3 | [export: 'my_custom_c_name']
      |  ^~~~~~
a.v:3:8: error: expected ']' before ':' token
    3 | [export: 'my_custom_c_name']
      |        ^
      |        ]
a.v:3:10: warning: character constant too long for its type        
    3 | [export: 'my_custom_c_name']
      |          ^~~~~~~~~~~~~~~~~~
b.c: In function 'main':
b.c:5:9: warning: implicit declaration of function 'my_custom_c_name' [-Wimplicit-function-declaration]
    5 |         my_custom_c_name();
      |         ^~~~~~~~~~~~~~~~
```

**Is it possible to be more clear about this section please ?**

**Comment by dumblob** [comment]
Without trying I guess the problem is `#include "a.v"` (which every C programmer knows does "plain text include" and therefore it is clear that the C compiler can not compile a `.v` file with *V* syntax inside :wink:). Try to include the by-V-compiler generated `.c` file and that shall work more or less.

**Reply by nicolasventer** [reply]
Thanks for the reply. What do you mean by : "include the by-V-compiler generated .c" file" ?
What is the command to get it ?

By the way, the information of "including the generated .c" should be in the section (and your next answer too) in order to allow any newcomer like me being able to call v from c.

**Reply by JalonSolov** [reply]
There are several ways to do what you want.  You could have V generate C code, then call it from your other C file.
```
v -o a.c a.v
gcc b.c a.c
```
Or you could create a shared lib with V then link your C code with it
```
v -shared a.v
gcc b.c a.so
```
etc.

**Reply by JalonSolov** [reply]
Just remove the `#include "a.v"` line from `b.c` before you try.  Depending on your C compiler, you may need to include a prototype in `b.c` for the function in `a.v`.

**Reply by nicolasventer** [reply]
I succeed to generate the C file with `v -o a.c a.v -shared`.
But when after I run `gcc -c a.c` I have the error:
``VERROR_MESSAGE Header file <gc.h>, needed for module `builtin` was not found. Please install the corresponding development headers.``.
How can I install the corresponding development headers ?

**Reply by JalonSolov** [reply]
Install the `libgc` package for your system.

**Reply by nicolasventer** [reply]
Thanks for the reply.
I executed on Linux (Ubuntu): `sudo apt-get install -y libgc-dev`
And then I compile with `gcc a.c b.c -lgc` and **it works!** 👍

**Comment by mnuccioarpae** [comment]
Try this.

a.v:
```v
module a

[export: 'my_custom_c_name']
fn foo() {
        println("foo")
}
```

b.c:
```v
void my_custom_c_name();

int main()
{
        my_custom_c_name();
}
```

```console
$ v -shared a.v
$ gcc b.c a.so
$ LD_LIBRARY_PATH=. ./a.out
foo
```

**Comment by nicolasventer** [comment]
Thanks everyone for all your replies.

I made a pull request that sum up the information you gave.


Here is the link : https://github.com/vlang/v/pull/15844

**Comment by 0xAvinash** [comment]
On my macbook I needed to use: `v -o some.c  -cross some.v` and compile the `some.c` file with `gcc -o some some.c` now you can run `./some` output file.

**Reply by 0xAvinash** [reply]
I think, I posted my solution in wrong thread.
